### PR TITLE
Feat: Restrict Access to Registration Form

### PR DESCRIPTION
**Pull request by NirmalHegde**
## Notion ticket link

[Ticket Name](https://www.notion.so/uwblueprintexecs/Waitlist-Restrict-access-to-registration-form-caf825d88fac4a83a04791c03914ea43)

## Implementation description
* Added restrictions for waitlisted applicants on the getCampById endpoint such that users cannot sign up for a camp when not available
* Added query parameters for waitlisted campers and camp session id and added checks to ensure that the link clicked is not expired

## Steps to test
1. Trigger endpoint /camp/{id}?wId={waitlistedCamperId}&session={campSessionId} and see if results returned are expected

## What should reviewers focus on?
* Ensure edge cases are accounted for

## Checklist
- [ ] My PR name is descriptive and in imperative tense
- [ ] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [ ] I have run the appropriate linter(s)
- [ ] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

File: backend/typescript/services/implementations/campService.ts [hunk 1]
```diff
@@ -125,6 +160,18 @@ class CampService implements ICampService {
...
+        (campSession) => campSession.id === campSessionId,
+      );
+
+      if (camp.campSessions.length === 0) {
+        throw new Error(
+          `No camp session matches with session ${campSessionId}`,
+        );
+      }
+    }
+
```
Comment (annechung) [inline backend/typescript/services/implementations/campService.ts:174]: can we also check the camp session's capacity here and throw an error if it's full? I think it's better than doing the check on the frontend, avoids doing a redirect? what do you think

Comment (NirmalHegde) [reply backend/typescript/services/implementations/campService.ts:174]: I think we agreed that the check would be done on the frontend, but yeah I can add that.

File: backend/typescript/services/implementations/campService.ts [hunk 2]
```diff
@@ -103,9 +106,41 @@ class CampService implements ICampService {
...
+      try {
+        const waitlistedCamper: WaitlistedCamper | null = await MgWaitlistedCamper.findById(
+          waitlistedCamperId,
+        );
+
+        if (!waitlistedCamper || !waitlistedCamper?.linkExpiry) {
+          throw new Error(
+            `Waitlisted Camper with Id ${waitlistedCamperId} does not exist or does not have an invite link.`,
+          );
+        }
```
Comment (annechung) [inline backend/typescript/services/implementations/campService.ts:125]: can we check here that `campSessionId == waitlistedCamper.campSession`?

File: backend/typescript/services/implementations/campService.ts [hunk 3]
```diff
@@ -103,9 +106,47 @@ class CampService implements ICampService {
...
 
+        if (campSessionId !== waitlistedCamper.campSession.toString()) {
+          throw new Error(
+            `Camp session ID ${campSessionId} does not exist or does not have an invite link.`,
```
Comment (annechung) [inline backend/typescript/services/implementations/campService.ts:128]: nit: better error msg?

```suggestion
            `Given waitlisted camper is not on the waitlist for given camp session.`,
```

File: backend/typescript/services/implementations/campService.ts [hunk 4]
```diff
@@ -103,9 +106,47 @@ class CampService implements ICampService {
...
 
+        if (campSessionId !== waitlistedCamper.campSession.toString()) {
+          throw new Error(
+            `Given waitlisted camper is not on the waitlist for given camp session.`,
+          );
+        }
+
+        const linkExpiryDate = new Date(waitlistedCamper.linkExpiry);
+        const currentDate = new Date();
+        if (linkExpiryDate > currentDate) {
```
Comment (annechung) [inline backend/typescript/services/implementations/campService.ts:134]: should be <=? 
```suggestion
        if (linkExpiryDate <= currentDate) {
```

Comment (NirmalHegde) [reply backend/typescript/services/implementations/campService.ts:134]: I think if they are registering on the current date, it should be allowed, but good catch

**Review by annechung — APPROVED** [review]
lgtm :D
